### PR TITLE
Add support for HassIO.

### PR DESCRIPTION
**Pull request by flz**
This will break existing configuration since the js script now reads
options.json (which is generated by hassio) as opposed to config.json.

### CODE REVIEW

**Comment by flz** [comment]
/data/options.json is where hassio will drop the configuration generated from the config.json template and the settings for the addon.

For people not using hassio, the file location isn't ideal. Not sure what's the best way to override it cleanly. Maybe an environment variable...

**Comment by tsightler** [comment]
I definitely don't want to merge an update that breaks existing deployments.  It would be pretty easy to update the code to have an environment variable override the default config file location so I'll consider that.

However, the main issue is that I personally don't have much interest in maintaining support for hassio, not that it's much effort from a code perspective, but it unfortunately increases the support burden due to the sheer number of questions, many from new users that don't fully understand the moving parts.  I already get 3-4 queries a week from such users saying the script doesn't work or whatever, when it's always just configuration, or not knowing how to use the GUI, or whatever.  I just don't think I have the time to service those users well enough.  For that reason, I'll have to think carefully on whether to merge this or not.

**Comment by broyuken** [comment]
Please reconsider adding hassio support. I'm fairly well versed in linux (debian/ubuntu based), but was looking to use hassio just to keep things as simple as possible. I tried to follow your instructions but I do not know how to get this running on alpine linux. I was able to clone the repo and npm install it, however I don't know how to create a system service for it, and I can't get the docker image to launch. There is another version based on the original script you referenced on the forums https://github.com/rs1932/ring-alarm-ha-integration

but it seems to be out of date, and right now it's not even installing to the hassio store because it is missing a required key. Yours seems to have more frequent updates and features and I would prefer to run it. I don't really care if I need to run it through hassio or manually install it on the box itself, I would just like it to work. I don't really want to spin up another vm to run this, but if that's the only solution I will have to.

**Comment by tsightler** [comment]
I appreciate the request, but this just doesn't make sense to me.  If you are fairly well versed in Linux, then running Home Assistant, either natively, or in a docker container, is so simple there's virtually no way that running it in hassio can be easier.  When I was considering support for hassio I spent far more time maintaining and fixing all of the brokeness in every hassio update than I ever did keeping my "production" Home Assistant install running.

The fact that rs1932's hassio wrapper for my script is broken, yet again (for at least the 3rd time so far this year) is exactly why I don't think I have time to support it.  I'm sure the fix is easy, but it still becomes my responsibility to fix it every time some arbitrary change in hassio breaks it and people expect me to fix it immediately because a new version came out yesterday.

That being said, I am still considering the possibility of officially supporting hassio when I migrate to the new version of the script which supports cameras, etc., so there's still hope that I might change my mind in the future, but the problem is, I don't use hassio, and would probably switch from Home Assistant to something else before I would choose to use it, so I have very little desire to maintain it.  I have to get over this psychological hump.

In the meantime, you're welcome to fork rs1932's work and maintain it yourself, or even submit a PR to his tree to update it.  It's nothing more than a simple wrapper that pulls down my script during install, so it stays updated with my updates.  That's the great thing about open source, the community can support and do whatever they are willing to put the work in to do, and the author (in this case, me) doesn't have to be responsible for maintaining things that they don't even use.

**Comment by broyuken** [comment]
I apologize, his wrapper isn't broken, I was trying to use another one. I actually found his a little while ago and it works fine. I'm very new to home assistant, moving from SmartThings and hassio seemed like the best choice to install. I like the fact that there's a "store" that I can install some things quickly such as mosquitto and the configuration editor in the sidebar. I'm assuming I could probably do that all on my own but this was a couple of clicks and I was up and running. As I get more involved in home assistant I may end up moving over to home assistant in a docker, probably once I mess around with docker a little more. 

All in all, I think my issue is resolved at this point because I was able to get it working. Your scripts are amazing and they work really well. The only thing I have to figure out is if a lock attached to the ring alarm would work with it, but that's some testing I can do on my own. Currently the lock is attached to SmartThings and communicating with home assistant via the smartthings bridge. 

Thank you again for all you have done, this is great!

**Comment by tsightler** [comment]
Locks "should" work, support for locks are in the code and others have reported success.

I do appreciate the feedback and support.  I will consider supporting hassio in the future, perhaps it is becoming more stable as they are pushing for a 1.0 release.  I just had a particularly frustrating experience with it and I much prefer to run "apt-get install mosquitto" and editing my config via "vi" vs installing from a store and dealing with whatever is broken (the mosquitto plugin went through a rough patch earlier this year).  It probably just put a bad taste in my mouth and, as a pretty experience admin, I generally find all these layers that make things "easier" to be much more complex to troubleshoot when things go wrong.  In hassio the code has to run in a docker container, so troubleshooting with a user becomes a pain compared to just being able to send them a new version of the script and having them restart.

Good luck and I will take your request seriously.

File: ring-alarm-mqtt.js [hunk 1]
```diff
@@ -423,7 +423,7 @@ function initMqtt() {
 
 // Get Configuration from file
 try {
-    CONFIG = require('./config')
+    CONFIG = require('/data/options')
```
Comment (KptnKMan) [inline ring-alarm-mqtt.js:426]:
```suggestion
    fs.access("/data/options.json", error => {
        if (!error) {
            // use hassio config, if found
            CONFIG = require('/data/options')
        } else {
            // use normal config
            CONFIG = require('./config')
        }
    });
```

I'm no no js expert, but might a fs check be more appropriate to not break compatibility?

**Review by KptnKMan — COMMENTED** [review]
Hi @tsightler and @flz noticed this PR and suggested something that might be a way to get around breaking compatibility.

I'm no js expert.

**Comment by tsightler** [comment]
This was actually my thought as well, but lack of time to work on this project has kept me from being able to test it.  Maybe I'll have some time over the next few weeks to do a little cleanup and pull in some of these but, unfortunately, this project is mostly on the back burner for now.

**Comment by respectTheCode** [comment]
I was thinking the same thing. You would want to do the check synchronously so that `CONFIG` has its value when the next line runs and uses `CONFIG`

```javascript
try {
  fs.accessSync('/data/options.json', fs.constants.R_OK);
  CONFIG = require('/data/options')
} catch (err) {
  CONFIG = require('./config')
}
```

https://nodejs.org/dist/latest-v12.x/docs/api/fs.html#fs_fs_accesssync_path_mode

**Comment by tsightler** [comment]
I've decided to keep Hass.io support in a separate tree for now.

https://github.com/tsightler/ring-mqtt-hassio-addon

**Pull request closed without merging**